Editor Media Google Photos: folders/albums view and dropdown

jakejones1984 opened the pull request:
## Description

Several issues have been raised about the possibility of being able to locate photos in external media via folders/albums, specifically Google Photos. This is within the context of the Media dialogue which appears within the Editor UI.

This PR will introduce a mechanism to allow users to select one of their folders from their Google Photos account and have the list of photos filter to show only photos within that folder.


## Features
The folder dropdown will have the following features:

- [x] ability to select from a dropdown of all your Google Photos folders
- [x] list of photos displayed will be filtered by the selected folder 
- [x] a list of Albums will be shown at the "root" level

## Todo

- [x] Move amends to `select-dropdown` into it's own PR (acting on advice from @ehg)

## Checklist:

- [x]  My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code follows the accessibility standards
- [x] My code has proper inline documentation.

jakejones1984 commented:
Copying over [what John Godley said](https://github.com/Automattic/wp-calypso/pull/25734#issuecomment-400217728):

> The last I looked the media library wasn't using redux, and so if you maintain folder state much like it is with search then the value can be passed to the request and passed along when a folder is chosen.
> 
> And yep, making an initial request to populate the dropdown sounds like it will be necessary.
> 
> I can't remember if albums are implemented in our API (or maybe they are untested), so we'll need to get that working too. They are present in Picasa so it shouldn't be a lot of work either way.

jakejones1984 commented:
> The last I looked the media library wasn't using redux, and so if you maintain folder state much like it is with search then the value can be passed to the request and passed along when a folder is chosen.

@johngodley Current folder is now passed into the query (see below screengrab from dev tools). The reference is the ID of the album in question.

<img width="171" alt="screen shot 2018-06-30 at 17 15 05" src="https://user-images.githubusercontent.com/40353716/42127011-4176f1ac-7c89-11e8-97a7-8b8d02100594.png">

Ccurrently this uses [fixture data](https://github.com/Automattic/wp-calypso/blob/17c4988c098cfd0ce34a49676ad515f6c1ae7983/client/my-sites/media-library/test/fixtures/index.js) based on the Picasa API. By default the folder is `all` which then needs be ignored by the WPCom and it should just retrieve all photos as it currently does.

jakejones1984 commented:
Update: @johngodley has suggested that the current Media implementation (on both Calypso and WPCom) isn't conducive to single folder based approach. As a result, I will be exploring an alternative approach in another PR. This implementation may still be useful in future.

a8ci18n commented on client/my-sites/media-library/media-folder-dropdown.jsx:
```diff
@@ -0,0 +1,102 @@
+		return [
+			{
+				value: '/',
+				label: this.props.translate( 'All Albums' ),
```
Hi! I've found a possible matching string that has already been translated **17** times:
`translate( 'Albums' )` *ES Score: 19*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

johngodley commented:
Pending a detailed review this all looks great. We will need a way to show recent photos as it’s currently known, and I think the media library should default to that view still.

Some questions:
- What happens to the dropdown if you have a lot of albums or no albums?
- How does the dropdown handle long album names?
- Does the refresh button also refresh the list of albums?
- Where is the album thumbnail coming from?
- What do we show for the thumbnail if the album is empty? (if we even show the album at all)

> I've also made some changes to allow the gallery to be keyboard navigable and to improve it's accessibility.

This seems like a good thing. Does it apply to the media library in general? If so it should be pushed out into a new PR

Also, I'm not sure if you've already tested these, but the media library is shown in other places:
- The 'media' option from the sidebar. There is an additional button in the subheader area when Google Photos is viewed here
- Settings > General  > Site icon > change

ehg commented:
This is looking good functionality wise. I like the album view, especially the Google auto-genned names:
<img width="1051" alt="screen shot 2018-07-09 at 14 45 14" src="https://user-images.githubusercontent.com/215074/42454196-c37e718a-8386-11e8-8bed-1defd26fa62f.png">

I have to seem to double-click to go into the albums?

Also, I got into a weird state whilst clicking around, leaving a blank screen. Switching albums/going back result in no XHR requests.

<img width="754" alt="screen shot 2018-07-09 at 14 39 39" src="https://user-images.githubusercontent.com/215074/42454055-6f2829dc-8386-11e8-8975-8c5bedf8f63c.png">

jakejones1984 commented:
> What happens to the dropdown if you have a lot of albums or no albums?
> How does the dropdown handle long album names?

Great point. Currently the DOM based dropdown will not provide a good experience. I recommend we move to a native `<select />` element as per the `FormSelect` component? This provides good a11y and ensures that all options can be accessed.

> Does the refresh button also refresh the list of albums?

Yes. It's the same functionality as the normal photos.

> Where is the album thumbnail coming from?

Every Album media object provided by Google has a `thumbnails` object with properties for various sizes. We utilise the most appropriate size.

johngodley commented:
Regarding single/double click I'm open either way. For me mobile maybe seems best for single, desktop for double. I don't think it should behave differently though.

> starting to need some design review

Always worth it! And not wanting to make you jump back and forth but some thoughts:

- Could 'Recent' be a folder in the dropdown?
- Do we even need the dropdown? If you can click via thumbnails and go back with a button then the dropdown isn't really needed. Of course you'd still need a way of toggling recent/folders

> I can still break this out if you'd prefer but I'm going to have to deliberately disable the linting checks on the lines in question to get this PR to pass

I'm not sure of the process for trials (Chris may be able to advise here) but it's a big PR so you'll probably need to break it up before deployment anyway. Your tests don't need to pass until the PR is ready, and when the a11y PR is deployed your tests will start working.

Another route is you could branch from your own branch. That is, the media library code could branch from the a11n fix branch. When the a11y fix goes in you can then rebase this PR and it will disappear.

Final point: it might be worth changing 'folders' to 'albums' as that's the term that Google uses

jakejones1984 commented:
Thanks @johngodley 

> Regarding single/double click I'm open either way.

On reflection we may revert back to single click as this seems more consistent across all devices (eg: you don't double-tap on mobile).

> Could 'Recent' be a folder in the dropdown?

I could and it would be trivial to implement. Will consider.

> Do we even need the dropdown?

Maybe not. Possibly useful if you want to avoid going back and forth between current folder and the folder list. 

> Final point: it might be worth changing 'folders' to 'albums' as that's the term that Google uses

I considered this, however what happens when we introduce another service who uses "folders"? I agree though, that for this release it might be worth making the UI elements refer to "Albums" for consistency. Most of it is configurable anyway, so when another service is introduce a different term can be utilised.

drw158 commented:
My design feedback/review. I'm not sure what should be a blocker or not since I don't know the scope or timeline of the project. I'll let y'all decide. If I feel strongly about something I'll let you know. Some of these problems exist already.

1. The initial view is not ideal because it skips some photos and is very limited on the max number it can show. The main problem is that this is not communicated in the UI, leaving the user to ask, "Where are my photos?". We should somehow show that this is just their "recent" photos, or find an alternative view, like the album view.
2. The transparent gallery icon on the albums should be removed.
3. It would be nice to vertically center the more [landscape images](https://user-images.githubusercontent.com/618551/42711342-228ee078-86ad-11e8-9198-a4da49b0f580.png) on the album thumbnails.
4. When you click into an album, select an image, and then go into another album, the image stays selected which is problematic if you don’t remember where the first image was. I think it should deselect the image if you navigate away.
5. The auto-generated albums, like `2017-04-14` are sort of distracting since I don't see those in Google Photos. Ultimately customers will probably just ignore them. Not a blocker.
6. If I select one photo, the primary button is 'insert'. For one video, it's `copy to media library`. More multiple items, it's `copy to media library`. As a user, I'm wondering why I'm copying to the media library when I explicitly clicked on the "Add" button in the post editor. I think all actions should be one or the other. If I choose just `insert` on a single image it seems to copy it to my media library anyway.
7. The toggle is a nice solution. I'd rather it be a segmented control since it's more about altering the view, not turning something on. The 2 text labels for the segmented control could be `Recent | Albums`. Or if you want to be more verbose, `Recent photos | Album view`.
8. I'd drop the folder icon next to the names in the album view.
9. The way you've implemented the back button is fine, but I'd rather see something more like this:

![screen shot 2018-07-13 at 4 51 27 pm](https://user-images.githubusercontent.com/618551/42715538-1479f6e8-86bd-11e8-9511-8d762499c7dd.png)

Once you click an album, it would switch to a new view with just the back button, without the other UI like the search bar and library dropdown. You can also see something similar in this mockup:

![screen shot 2018-07-13 at 4 55 31 pm](https://user-images.githubusercontent.com/618551/42715668-9b5056e4-86bd-11e8-8a36-bd3b3237cc45.png)


Generally I'm having trouble with the functionality working, but I think the live calypso branch is the problem. Images aren't loading, etc.

johngodley commented:
Regarding point 6, this is a technical limitation that isn't related to this PR.

jakejones1984 commented:
Firstly, thank you @drw158 for taking the time to review this. I really appreciate it. Feedback below:

> The initial view is not ideal because it skips some photos and is very limited on the max number it can show. The main problem is that this is not communicated in the UI, leaving the user to ask, "Where are my photos?". We should somehow show that this is just their "recent" photos, or find an alternative view, like the album view.

Understood. Unfortunately however, this has _already_ been reverted back to what it is now from what you are suggesting (ie: "just show Albums view" - which is what I implemented initially).

 I should say that there is also no clear agreement that the current Google Photos view is actually "Recent". I believe that it is, but @johngodley (who has much more experience in this area having written much of the code) believes that it is not. 

Either way the key issue is that there _is_ a hard limit, so whether or not it is officially "Recent" the net result is that it only shows a limited number of posts. Therefore I've elected to leave the default view as "Recent". It is my opinion that to avoid bloating this PR, the issue of the "Recent" not displaying a clear warning on the `300` photo limited needs to be tackled in a separate PR.

> The transparent gallery icon on the albums should be removed.

Removed as discussed.

> It would be nice to vertically center the more landscape images on the album thumbnails.

Good point. Generally I felt this was looking a bit scrappy even when I vertically aligned it. Therefore I've elected to always have the image "fill" the available space. This does result in cropping but as it's only a preview I feel it's an acceptable compromise to avoid the "messy" UI that is generated when you have a Google Photos Album list that has a large variety of Album cover image ratios.

<img width="764" alt="screen shot 2018-07-14 at 17 21 07" src="https://user-images.githubusercontent.com/40353716/42727807-2daea882-87a5-11e8-83d2-98253aa5655d.png">

> When you click into an album, select an image, and then go into another album, the image stays selected which is problematic if you don’t remember where the first image was. I think it should deselect the image if you navigate away.

I've added this functionality but I'd like to check with @ehg and @johngodley that this isn't intended functionality.

![media_clear_selected](https://user-images.githubusercontent.com/40353716/42727819-42b42f72-87a5-11e8-81ac-98ec3bd34d96.gif)

> The auto-generated albums, like 2017-04-14 are sort of distracting since I don't see those in Google Photos. Ultimately customers will probably just ignore them. Not a blocker.

Understood. We get this from the API so we can't control this. Interestingly i've had mixed feedback. For example, @ehg actually liked this feature. Glad you agree it's not a blocker. 

> If I select one photo, the primary button is 'insert'. For one video, it's copy to media library. More multiple items, it's copy to media library. As a user, I'm wondering why I'm copying to the media library when I explicitly clicked on the "Add" button in the post editor. I think all actions should be one or the other. If I choose just insert on a single image it seems to copy it to my media library anyway.

As noted by @johngodley this is a technical limitation. It would need to be addressed as a separate PR.

> The toggle is a nice solution. I'd rather it be a segmented control since it's more about altering the view, not turning something on. The 2 text labels for the segmented control could be Recent | Albums. Or if you want to be more verbose, Recent photos | Album view.

I've updated this as you suggested with the shorter text form and a `title` attribute containing the longer more descriptive form of the wording.

<img width="771" alt="screen shot 2018-07-14 at 17 39 01" src="https://user-images.githubusercontent.com/40353716/42727839-8318aad4-87a5-11e8-9765-a98b7e4a2213.png">

> I'd drop the folder icon next to the names in the album view.

Done. I added this because I wasn't confident it was clear enough they were Albums. I'll defer to you on this.

> The way you've implemented the back button is fine

That's good to hear! 👍 

> but I'd rather see something more like this

What you propose sounds (and looks) interesting but won't be simple to implement and will require a separate PR. Can I confirm that you don't see this as a blocker. Does what we have work as an MVP?

johngodley commented:
> We should somehow show that this is just their "recent" photos, or find an alternative view, like the album view.

Although the current recent view is now much more restricted than it used to be, I’d still like it to remain as the default, with albums being an addition for those that use them. I think it’s the view that people expect to see by default, and the one that will likely fit most usage patterns (especially if we get Google Photos into the mobile app). I'm not sure people will be confused - this hasn't really happened so far, and with the album UI it's now more explicit.

Once we switch to the new Google API the recent view will function exactly as expected. In the mean time showing the ‘no more photos can be shown’ message at 300 rather than 1000 photos is a good compromise. It’s fine for us to change the message to indicate this is caused by external factors.

> I've added this functionality but I'd like to check with @ehg and @johngodley that this isn't intended functionality.

I do recall the media library remembering selected image in some situations, and this is more of an accident - the media library code was never designed for all the jumping about. If it’s possible to clear the current selection when changing album then it should probably do that.

jakejones1984 commented:
> In the mean time showing the ‘no more photos can be shown’ message at 300 rather than 1000 photos is a good compromise. It’s fine for us to change the message to indicate this is caused by external factors.

@johngodley Please see [this PR](https://github.com/Automattic/wp-calypso/pull/26129) for an implementation.

> If it’s possible to clear the current selection when changing album then it should probably do that.

Great. That has been implemented.

ehg commented on client/my-sites/media-library/media-folder-dropdown.jsx:
```diff
@@ -0,0 +1,119 @@
+	renderOptions( optionData ) {
+		return optionData.map( folder => {
+			const folderId = '' + folder.ID;
+			const folderChildren = folder.children ? ` (${ folder.children })` : ``;
```
Nitpick, we can get away without a template literal for the empty string here.

Also, can a folder's children have children?

jakejones1984 replied:
Fixed string in [this commit](https://github.com/Automattic/wp-calypso/pull/25763/commits/14cc2b93c528633ae2fbbe1507b3dff09382ba41)


> can a folder's children have children?

A folders children could have children but for MVP Google Photos implementation it only ever has 1 level of folders.

ehg commented on client/my-sites/media-library/media-folder-dropdown.jsx:
```diff
@@ -0,0 +1,119 @@
+		return (
+			<div className={ rootClassNames }>
+				<FormLabel htmlFor="media-library-folders" className="media-library__folder-dropdown-label">
+					Select Folder
```
Should this be translated?

jakejones1984 replied:
Yes. Fixed as of [this commit](https://github.com/Automattic/wp-calypso/pull/25763/commits/f64420e779546eefdaacf654df1a6d23c27f44a2)

ehg commented on client/my-sites/media-library/content.jsx:
```diff
@@ -59,6 +60,40 @@ class MediaLibraryContent extends React.Component {
+
+		this.state = { folders: [] };
+
+		this.parseFolders = this.parseFolders.bind( this );
```
Let's get rid of these `bind`s and change the function definitions, e.g. `parseFolders = ( { media } ) => {`

jakejones1984 replied:
Fixed as of [this commit](https://github.com/Automattic/wp-calypso/pull/25763/commits/bb404944bde001be1789cbdccebf1ebc811c223f)

jakejones1984 commented on client/my-sites/media-library/external-media-header.jsx:
```diff
@@ -123,23 +150,104 @@ class MediaLibraryExternalHeader extends React.Component {
+		}.bind( this );
+	}
+
 	renderCard() {
```
Agreed - this was getting far too large and complex. It has now been broken up into smaller render functions.

a8ci18n commented on client/my-sites/media-library/external-media-header.jsx:
```diff
@@ -123,28 +159,148 @@ class MediaLibraryExternalHeader extends React.Component {
+				folder={ this.props.folder }
+				defaultOption={ {
+					ID: '/',
+					name: this.props.translate( 'All Albums' ),
```
Hi! I've found a possible matching string that has already been translated **17** times:
`translate( 'Albums' )` *ES Score: 19*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

jakejones1984 replied:
@ehg Looks like `Albums` has already been translated. Is it best to

1. Call `translate()` twice - once for `All` and again for `Albums`?
2. Utilise [String Substitution](https://github.com/Automattic/i18n-calypso#string-substitution)?
3. Leave "as is"?

Appreciate your advice here.

ehg replied:
I think it's fine as is.

getdave commented:
Still intending to land this but waiting on changes discussed in `p4hfux-4s0-p2#comment-6708`.
